nifi-commons/nifi-utils/src/main/java/org/apache/nifi/retry/NoReturnCallable.java

```diff
@@ -14,25 +14,16 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.nifi.controller.status.history.storage;
-
-import org.apache.nifi.controller.status.history.GarbageCollectionHistory;
-import org.apache.nifi.controller.status.history.GarbageCollectionStatus;
-
-import java.time.Instant;
+package org.apache.nifi.retry;
 
 /**
- * Readable status storage for garbage collection status entries.
+ * Represents a portion of callable code without expected return value.
  */
-public interface GarbageCollectionStatusStorage extends StatusStorage<GarbageCollectionStatus> {
+@FunctionalInterface
+public interface NoReturnCallable {
```

Comment:
I think, from Java Gerenics standpoint, `Void` could work out. Where it fails however is the way of intended usage. The `RetryTemplate` is primarily designed to be called not with explicit implementations of `Callable` or `NoReturnCallable` for that sake, but to give lambda expressions as arguments, like the following (from `SynchronousRetryTemplateTest`):

`testSubject.execute(() -> action.getValue("okay"), FALLBACK_ACTION);`

Creating an expression with `Void` return value to fit this intended use look to be cumbersome. So by adding `NoReturnCallable` as a functional interface, which internally wraps these "void return" or "no return" lambdas  into a 'Callable' will solve the question of easy use. With this, a no return expression can be as simple as:

`testSubject.executeWithoutValue(() -> action.consumeValue("okay"));`

Reply:
Thanks for the additional background. With that in mind, it seems like having the custom `Callable` as a nested interface on the RetryTemplate would be a better approach.

Did you consider using [spring-retry](https://www.baeldung.com/spring-retry#retrytemplate) instead? Since the abstraction applies to QuestDB, it seems worth considering as it has a minimal number of dependencies.

Reply:
Good idea, I tried and it looks to me that the sprint-retry can be configured to provide the same behaviour.